src/compose/images.ts

```diff
@@ -398,35 +434,75 @@ async function getImagesForCleanup(): Promise<string[]> {
-): Promise<Docker.ImageInspectInfo> {
-	try {
-		const image = await docker.getImage(imageName);
-		return await image.inspect();
-	} catch (e) {
-		if (NotFoundError(e)) {
-			const digest = imageName.split('@')[1];
-			let imagesFromDb: Image[];
-			if (digest != null) {
-				imagesFromDb = await db
-					.models('image')
-					.where('name', 'like', `%@${digest}`);
-			} else {
-				imagesFromDb = await db
-					.models('image')
-					.where({ name: imageName })
-					.select();
-			}
+export async function inspectByName(imageName: string) {
+	// Fail fast
+	assert(!!imageName, 'image name to inspect cannot be null');
 
-			for (const image of imagesFromDb) {
-				if (image.dockerImageId != null) {
-					return await docker.getImage(image.dockerImageId).inspect();
-				}
-			}
-		}
-		throw e;
-	}
+	// Get image by the full image URI. This will only work for regular pulls
+	// and old style images `repo:tag`
+	const inspectByURI = async () => await docker.getImage(imageName).inspect();
+
+	const {
+		registry,
+		imageName: name,
+		tagName,
+		digest,
+	} = dockerUtils.getRegistryAndName(imageName);
+
+	// Look for an image in the engine with registry/image as reference (tag)
+	// for images with deltas this should return unless there is some inconsistency
+	// and the tag was deleted
+	const repo = [registry, name].filter((s) => !!s).join('/');
+	const reference = [repo, tagName].filter((s) => !!s).join(':');
+
+	const inspectByReference = async () =>
+		await docker
+			.listImages({
+				digests: true,
+				filters: { reference: [reference] },
+			})
+			.then(([img]) =>
+				!!img
+					? docker.getImage(img.Id).inspect()
+					: Promise.reject(
+							new StatusError(
+								404,
+								`Failed to find an image matching ${imageName}`,
+							),
+					  ),
+			);
+
+	// Otherwise look in the database for an image with same digest or same name and
+	// get the dockerImageId from there. If this fails the image may still be on the
+	// engine but we need to re-trigger fetch and let the engine tell us if the
+	// image data is there
+	const inspectByDigest = async () =>
+		await db
+			.models('image')
+			.where('name', 'like', `%${digest}`)
+			.orWhere({ name: imageName }) // Default to looking for the full image name
+			.select()
+			.then((images) =>
+				images.filter((img: Image) => img.dockerImageId !== null),
+			)
+			// Assume that all db entries will point to the same dockerImageId, so use
+			// the first one. If this assumption is false, there is a bug with cleanup
+			.then(([img]) =>
+				!!img
+					? docker.getImage(img.dockerImageId).inspect()
+					: Promise.reject(
+							new StatusError(
+								404,
+								`Failed to find an image matching ${imageName}`,
+							),
+					  ),
+			);
+
+	// Convert the promise array into a promise chain
+	return await [inspectByURI, inspectByReference, inspectByDigest].reduce(
+		(res, p) => res.catch(() => p()),
+		Promise.reject('You should never see this'),
```

Comment:
The message is a placeholder really, unless someone breaks the implementation it will never make it to the user. I think the message should reflect that idea, although I don't necessarily think my current message reflects that well

Reply:
I think is the same as with the assertion on top of the method, if it does fail we want it to fail in a way that ensures we catch it before it can get to production. It really won't fail since we are using a static array as source for the `reduce`

Since it's causing confusing let me see if I can find a way to perform the `reduce` without passing an initial value. I couldn't think of a clean way to do it but I can give it another try

Reply:
Opted to update the reject message with something that makes it totally obvious that is a bug